### PR TITLE
Adding support of markdown exported

### PR DESCRIPTION
**Pull request by Anipik**
Markdown Exporter in asmdiff will allows us to generate api-diffs on the fly.
Most of the code has been taken from https://devdiv.visualstudio.com/DevDiv/_git/CoreFxTools?path=%2Fsrc%2Fapps%2FApiReviewer

This was a winforms app. The calling code is the new one.

### CODE REVIEW

File: src/Microsoft.DotNet.AsmDiff/ApiRecordingCSharpDiffWriter.cs [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+        }
+
+        private DiffStyle GetCurrentDiffStyle()
+        {
+            return _styleStack.Aggregate(DiffStyle.None, (current, diffStyle) => current | diffStyle);
+        }
+
+        public void Write(string str)
+        {
+            // HACK: Work around issue where member attributes are emitted
```
Comment (ericstj) [inline src/Microsoft.DotNet.AsmDiff/ApiRecordingCSharpDiffWriter.cs:209]: Please avoid the use of `HACK`.  Typically policheck will flag this.

**Comment by ericstj** [comment]
Is this a copy of existing code or is it net-new code?  If a copy, where did it come from?  Leave a breadcrumb here (even if to an internal repro).

File: src/Microsoft.DotNet.AsmDiff/ApiRecordingCSharpDiffWriter.cs [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+            return representative.Name.Value == "value__";
+        }
+
+        private static bool IsDelegateMember(ITypeDefinitionMember representative)
+        {
+            return representative.ContainingTypeDefinition.IsDelegate;
+        }
+    }
+
+    internal sealed class DiffRecorder : IStyleSyntaxWriter
```
Comment (ericstj) [inline src/Microsoft.DotNet.AsmDiff/ApiRecordingCSharpDiffWriter.cs:153]: Should this use 1 class per file?

File: src/Microsoft.DotNet.AsmDiff/ApiRecordingCSharpDiffWriter.cs [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+                    return DiffStyle.NotCompatible;
+                default:
+                    throw new ArgumentOutOfRangeException("style");
+            }
+        }
+    }
+    
+    public static class DiffExtensions
+    {
+        public static bool HasStyle(this DiffToken token, DiffStyle diffStyle)
```
Comment (ericstj) [inline src/Microsoft.DotNet.AsmDiff/ApiRecordingCSharpDiffWriter.cs:313]: This extension method appears to have a single call-site.  Perhaps we should just move this closer to that callsite: https://github.com/dotnet/arcade/pull/5571/files#diff-8c2954f01a8846abf132bb0f32ec9f51R273

File: src/Microsoft.DotNet.AsmDiff/ApiRecordingCSharpDiffWriter.cs [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+        {
+            // Special case the zero-flag.
+            if (diffStyle == DiffStyle.None)
+                return token.Style == DiffStyle.None;
+
+            return (token.Style & diffStyle) == diffStyle;
+        }
+    }
+
+    internal sealed class DisposeAction : IDisposable
```
Comment (ericstj) [inline src/Microsoft.DotNet.AsmDiff/ApiRecordingCSharpDiffWriter.cs:323]: Looks to be a duplicate of https://github.com/dotnet/arcade/blob/04f4b34837651ad724f48362285f0a5d2ffa4a90/src/Microsoft.Cci.Extensions/Internal/DisposeAction.cs.  Which makes me wonder: should we put this with the other IStyleSyntaxWriter in CCI.Extensions?

File: src/Microsoft.DotNet.AsmDiff/ApiRecordingCSharpDiffWriter.cs (hunk 2 above)
Comment (Anipik) [reply src/Microsoft.DotNet.AsmDiff/ApiRecordingCSharpDiffWriter.cs:153]: I combined the classes which are used together eg diff recorder is only used in ApiRecordingCSharpDiffWriter.

File: src/Microsoft.DotNet.AsmDiff/ApiRecordingCSharpDiffWriter.cs (hunk 4 above)
Comment (Anipik) [reply src/Microsoft.DotNet.AsmDiff/ApiRecordingCSharpDiffWriter.cs:323]: We can move DiffRecorder but we will have to move some of the "markdown" specific classes to the cci as well.

File: src/Microsoft.DotNet.AsmDiff/ApiRecordingCSharpDiffWriter.cs (hunk 2 above)
Comment (ericstj) [reply src/Microsoft.DotNet.AsmDiff/ApiRecordingCSharpDiffWriter.cs:153]: Still it should follow one-class-per-file convention, unless it is a nested class.

File: src/Microsoft.DotNet.AsmDiff/ApiRecordingCSharpDiffWriter.cs (hunk 4 above)
Comment (ericstj) [reply src/Microsoft.DotNet.AsmDiff/ApiRecordingCSharpDiffWriter.cs:309]: If you don't want to move, can you make it common source?  EG: move DisposeAction.cs under src/Common

File: src/Microsoft.DotNet.AsmDiff/DiffApiDefinition.cs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+            Children = new ReadOnlyCollection<DiffApiDefinition>(children);
+        }
+
+        public override string ToString()
+        {
+            return Difference.ToString().Substring(0, 1) + " " + Definition.UniqueId();
+        }
+    }
+
+    internal static class NameHelper
```
Comment (ericstj) [inline src/Microsoft.DotNet.AsmDiff/DiffApiDefinition.cs:42]: Similar feedback here.  If this is only used here, just combine these as private static members on DiffApiDefinition

**Pull request merged**